### DISCUSSION TITLE
Animation - Ability to blend property and bezier tracks

### DISCUSSION THREAD

**Discussion by Arnklit** [godotengine/godot-proposals#10175] [General Discussions]
It is currently not possible to blend regular property tracks and bezier track animations. see https://docs.godotengine.org/en/stable/tutorials/animation/animation_track_types.html#bezier-curve-track - as far as I can tell @TokageItLab added the note here, so I'm guessing they'll know the details.

It is really suboptimal that we allow for the different types of animation tracks, but that they don't mix together. Users will create different track types depending on what's easiest in a given situation and not realise until later that they don't mix and they've created a lot more work for themselves having to convert half their animations.

Would it be possible to get blending of these two types working? What are the technical reasons that it doesn't work at the moment?

If it's not possible, should Animation libraries only allow for one type of tracks so people avoid this issue?

**Comment by TokageItLab** [comment]
The main reason is that the structure of the paths is different as I wrote sometimes: For example, a trackpath `player:position:x` does not determine whether "x" is 2D or 3D, or whether it is a custom property made by user, so it is necessary to actually search the scene to get the correct VariantType. - The BezierTrack and ValueTrack were added a long time ago, before I committed to Godot, and was not well designed at the time.

ValueTrack uses the path `player:position`, while BezierTrack uses the path `player:position:x`. Currently (after 4.3.dev, by https://github.com/godotengine/godot/pull/86687), you can blend them now by manually rewriting the ValueTrack path to `player:position:x`, but it is inconvenient.

Because BezierTrack **splits elements when tracks are added**, finding the corresponding ValueTrack requires finding the properties of the element before splitting from the actual object.

Conversely, finding the corresponding BezierTrack from a ValueTrack requires dynamically splitting the element. Both are ridiculous from a performance standpoint.

I think the proper direction is to eliminate the Bezier track and integrate it into the Value track. Since the Bezier interpolation has poor performance since it do interpolate Beziers in playbacking. It should only be done during editing and use baked values depending on FPS in runtime. In other words, the current BezierTrack data should be editor only.

In that case (old BezierTrack -> new ValueTrack which support Bezier), the converter must be needed. So I don't think this can be accomplished in Godot 4, and if Godot 5 comes out, this is the main thing I would like to work on there.
